casper/protocols/sharding/sharding_view.py

```diff
@@ -0,0 +1,152 @@
+    """A view class that keeps track of children, latest_messages_on_shard and starting_blocks"""
+    def __init__(self, messages=None, shard_genesis_block=None):
+        self.children = dict()
+
+        self.shard_genesis_blocks = dict()  # shard_id -> genesis for shard
+        self.starting_blocks = dict()  # shard_id -> starting block for forkchoice
+        self.latest_messages_on_shard = dict()  # shard_id -> validator -> message
+
+        self.select_shards = self.select_random_shards
+
+        if shard_genesis_block:
+            for shard_id in shard_genesis_block.estimate['shard_ids']:
+                self.shard_genesis_blocks[shard_id] = shard_genesis_block
+                self.starting_blocks[shard_id] = shard_genesis_block
+
+        super().__init__(messages)
+
+    def estimate(self):
+        """Returns the current forkchoice in this view"""
+        shards_forkchoice = dict()
+
+        for shard_id in sorted(self.starting_blocks):
+            shard_tip = forkchoice.get_shard_fork_choice(
+                self.starting_blocks[shard_id],
+                self.children,
+                self.latest_messages_on_shard[shard_id],
+                shard_id
+            )
+            shards_forkchoice[shard_id] = shard_tip
+
+            left_child_shard = shard_id + '0'
+            right_child_shard = shard_id + '1'
+            if left_child_shard in self.starting_blocks:
+                self.set_child_starting_block(shard_tip, shard_id, left_child_shard)
+            if right_child_shard in self.starting_blocks:
+                self.set_child_starting_block(shard_tip, shard_id, right_child_shard)
+
+        self.check_forkchoice_atomicity(shards_forkchoice)
+
+        shards_to_build_on = self.select_shards(shards_forkchoice)
+        return {'prev_blocks': {shards_forkchoice[shard_id] for shard_id in shards_to_build_on},
+                'shard_ids': shards_to_build_on}
+
+    def set_child_starting_block(self, tip_block, parent_id, child_id):
+        """Changes the starting block for the forkchoice of a shard"""
+        child_merge_block = self.previous_merge_block_on_shard(
+            tip_block,
+            parent_id,
+            child_id
+        )
+
+        if child_merge_block:
+            self.starting_blocks[child_id] = child_merge_block
+        else:
+            self.starting_blocks[child_id] = self.shard_genesis_blocks[child_id]
+
+    def select_random_shards(self, shards_forkchoice):
+        """Randomly selects a shard to build on, and sometimes selects another child shard"""
+        shards_to_build_on = [r.choice([key for key in self.starting_blocks.keys()])]
+        if (r.randint(0, 1) == 1):
+            child = str(r.randint(0, 1))
+            if shards_to_build_on[0] + child in self.starting_blocks:
+                shards_to_build_on.append(shards_to_build_on[0] + child)
+
+        return set(shards_to_build_on)
+
+    def check_forkchoice_atomicity(self, shards_forkchoice):
```

Comment:
Okay here right now, but might be more appropriate in some sort of testing utils when we add testing for sharding.